code/net/ConnectX-5/mlx5.c

```diff
@@ -0,0 +1,412 @@
+// All references in this file are to the Mellanox ConnectX-5 Datasheet unless otherwise noted.
+// It can be found at TODO
+
+#include "net/network.h"
+
+#include "env/endian.h"
+#include "env/memory.h"
+#include "env/time.h"
+#include "util/log.h"
+
+#ifndef __cplusplus
+// Don't include <assert.h> since that's not allowed in freestanding implementations
+#define static_assert _Static_assert
+#endif
+
+// The PCIREG_ values are register indexes.
+// The REG_ values are register indexes, some of which take as argument an index.
+// The sub-values are fields, which can be one or multiple bits.
+// The following two macros make defining fields easier; note that bit indices start at 0.
+#define BIT(n) (1u << (n))
+#define BITS(start, end) (((end) == 31 ? 0u : (0xFFFFFFFFu << ((end) + 1))) ^ (0xFFFFFFFFu << (start)))
+// This is for bits when we need 64-bit numbers
+#define BITL(n) (1ull << (n))
+
+#define PCIREG_ID 0x00u
+// https://en.wikipedia.org/wiki/PCI_configuration_space#/media/File:Pci-config-space.svg
+// PCIe Configuration Space Summary: "0x10 Base Address Register 0" (32 bit), "0x14 Base Address Register 1" (32 bit)
+#define PCIREG_BAR0_LOW 0x10u
+#define PCIREG_BAR0_HIGH 0x14u
+
+#define PCIREG_COMMAND 0x04u
+#define PCIREG_COMMAND_MEMORY_ACCESS_ENABLE BIT(1)
+#define PCIREG_COMMAND_BUS_MASTER_ENABLE BIT(2)
+#define PCIREG_COMMAND_INTERRUPT_DISABLE BIT(10)
+
+// Driver constants
+#define PCI_ID_HIGH 0x1017u
+#define PCI_ID_LOW 0x15b3u
+
+#define FW_REV_MINOR 0x1000
```

Comment:
It'd be good to write down where you got those - did you read them from the card or are they somewhere in the data sheet?